django/test/testcases.py

```diff
@@ -1119,12 +1173,17 @@ def setUpClass(cls):
                     cls._rollback_atomics(cls.cls_atomics)
                     cls._remove_databases_failures()
                     raise
+        pre_attrs = set(cls.__dict__)
         try:
             cls.setUpTestData()
         except Exception:
             cls._rollback_atomics(cls.cls_atomics)
             cls._remove_databases_failures()
             raise
+        new_attrs = set(cls.__dict__) - pre_attrs
+        for name in new_attrs:
+            value = getattr(cls, name)
+            setattr(cls, name, TestData(name, value))
```

Comment:
Here's a thought: this technique won't detect..

```python
class MyTests(TestCase):
    foo = None
    @classmethod
    def setUpTestData(cls):
        foo = Foo.objects.create(...)
```

I find people coming from other languages like Java often think they need to put `foo = None` or similar in the class to reserve space for it.

Maybe better to detect if attribute names OR values have changed?

Reply:
(I was going to suggest proxying `cls.__dict__` temporarily, but a class's dict can't be swapped.)

Reply:
yeah good catch. I can think of three ways of handling that case.

1. Make `TestCase` use a metaclass that overrides `__setattr__` to behave differently during `setUpTestData`.
2. Make `TestCase` use a metaclass that overrides `__prepare__` to return a mapping that behaves differently during `setUpTestData`.
3. Store a copy of `cls.__dict__` and use identity checks (`is` operation) to compare what changed during `setUpTestData`.

1. and 2. are clever but 3. is way less invasive and should work so I'll focus on that.